pallets/asset/src/lib.rs

```diff
@@ -2585,19 +2608,56 @@ impl<T: Config> Module<T> {
         Self::ensure_asset_exists(&ticker)?;
         Self::ensure_asset_type_valid(asset_type)?;
         let did = <ExternalAgents<T>>::ensure_perms(origin, ticker)?;
-
-        Tokens::mutate(ticker, |token| token.asset_type = asset_type);
+        Tokens::try_mutate(&ticker, |token| -> DispatchResult {
+            Self::ensure_valid_asset_type_update(token.asset_type, asset_type)?;
+            token.asset_type = asset_type;
+            Ok(())
+        })?;
         Self::deposit_event(RawEvent::AssetTypeChanged(did, ticker, asset_type));
         Ok(())
     }
 
     /// Returns `None` if there's no asset associated to the given ticker,
-    /// returns Some(true) if the asset exists and is of type `AssetType::NFT`, and returns Some(false) otherwise.
+    /// returns Some(true) if the asset exists and is of type `AssetType::NonFungible`, and returns Some(false) otherwise.
     pub fn nft_asset(ticker: &Ticker) -> Option<bool> {
         let security_token = Tokens::try_get(ticker).ok()?;
         match security_token.asset_type {
-            AssetType::NFT => Some(true),
+            AssetType::NonFungible(_) => Some(true),
             _ => Some(false),
         }
     }
+
+    /// Returns Ok if both parameters are non fungible types or if both are fungible types.
+    /// Otherwise it returns an error.
+    fn ensure_valid_asset_type_update(
+        current_asset_type: AssetType,
+        new_asset_type: AssetType,
+    ) -> DispatchResult {
+        match current_asset_type {
+            AssetType::NonFungible(_) => {
+                // If the current asset type is non fungible, it cannot be updated to a fungible type.
+                if let AssetType::NonFungible(_) = new_asset_type {
+                    return Ok(());
+                }
+                Err(Error::<T>::IncompatibleAssetTypeUpdate.into())
+            }
+            AssetType::EquityCommon
+            | AssetType::EquityPreferred
+            | AssetType::Commodity
+            | AssetType::FixedIncome
+            | AssetType::REIT
+            | AssetType::Fund
+            | AssetType::RevenueShareAgreement
+            | AssetType::StructuredProduct
+            | AssetType::Derivative
+            | AssetType::Custom(_)
+            | AssetType::StableCoin => {
+                // If the current asset type is fungible, it cannot be updated to a non fungible type.
+                if let AssetType::NonFungible(_) = new_asset_type {
+                    return Err(Error::<T>::IncompatibleAssetTypeUpdate.into());
+                }
+                Ok(())
+            }
+        }
```

Comment:
A `is_fungible()` method on `AssetType` could be used to simplify the logic here.  And be used in the other NFT checks.

```suggestion
        if current_asset_type.is_fungible() != new_asset_type.is_fungible() {
            return Err(Error::<T>::IncompatibleAssetTypeUpdate.into());
        }
       Ok(())
```

Reply:
That's better. I created both `is_fungible` and `is_non_fungible` just to make sure that we are forced to reassess things in case some like "semi-fungible" comes up in the future. Otherwise, `ensure!(!asset_type.is_fungible())` would allow this new type to create an nft collection, which we might not want to.

Reply:
If we add a new top-level AssetType (i.e. `AssetType::SemiFungible`), then we might want to have another enum for just the variants without the sub type.  Like what we do with the `Claim` and `ClaimType` enums.

But for now having `is_fungible` and `is_non_fungible` is good and easy to update when adding more top-level types.